pkg/controller/migplan/registry_health.go

```diff
@@ -0,0 +1,144 @@
+type registryHealth struct {
+	hostClient client.Client
+	Interval   time.Duration
+	handler    handler.EventHandler
+	queue      workqueue.RateLimitingInterface
+	predicates []predicate.Predicate
+	podLabels  map[string]string
+	planLabels map[string]string
+}
+
+// Start the health checks
+func (r *registryHealth) Start(
+	handler handler.EventHandler,
+	queue workqueue.RateLimitingInterface,
+	predicates ...predicate.Predicate) error {
+
+	r.handler = handler
+	r.queue = queue
+	r.predicates = predicates
+	go r.run()
+
+	return nil
+}
+
+// Enqueue a reconcile request event for migplan
+func (r *registryHealth) enqueue(plan v1alpha1.MigPlan) {
+	event := event.GenericEvent{
+		Meta:   &plan.ObjectMeta,
+		Object: &plan,
+	}
+	for _, predicate := range r.predicates {
+		if !predicate.Generic(event) {
+			return
+		}
+	}
+
+	r.handler.Generic(event, r.queue)
+}
+
+//Run the health checks for registry pods
+func (r *registryHealth) run() {
+	//List all the migplans that are in running state using the hostClient
+	//Now using srcClient and destClient for each migplan find the registry pods, if registry container is not ready then enqueue this migplan
+	//repeat all the above steps for all the plans for both clusters
+
+	for {
+		time.Sleep(r.Interval)
+		planList, err := v1alpha1.ListPlansWithLabels(r.hostClient, r.planLabels)
+		if err != nil {
+			log.Trace(err)
+			return
+		}
+		if planList != nil {
+			for _, plan := range planList {
+				//TODO avoid race condition check
+				srcCluster, err := plan.GetSourceCluster(r.hostClient)
+				if err != nil {
+					log.Trace(err)
+					//TODO Error condition would result in loosing the watch util the pod restarts
+				}
+
+				if !srcCluster.Status.IsReady() {
+					log.Info("Cannot check registry pod health, cluster is not ready", srcCluster.Name, plan.Name)
+				}
+
+				srcClient, err := srcCluster.GetClient(r.hostClient)
+				if err != nil {
+					log.Trace(err)
+				}
+
+				destCluster, err := plan.GetDestinationCluster(r.hostClient)
+				if err != nil {
+					log.Trace(err)
+				}
+
+				if !destCluster.Status.IsReady() {
+					log.Info("Cannot check registry pod health, cluster is not ready", destCluster.Name, plan.Name)
+				}
+
+				destClient, err := destCluster.GetClient(r.hostClient)
+				if err != nil {
+					log.Trace(err)
+				}
+
+				srcRegistryPods, err := getRegistryPods(&plan, srcClient)
+				if err != nil {
+					log.Trace(err)
+				}
+
+				destRegistryPods, err := getRegistryPods(&plan, destClient)
+				if err != nil {
+					log.Trace(err)
+				}
+
+				if r.checkPodHealthAndPlanConditionToEnqueue(srcRegistryPods, &plan) {
+					r.enqueue(plan)
+					continue
+				}
+
+				if r.checkPodHealthAndPlanConditionToEnqueue(destRegistryPods, &plan) {
+					r.enqueue(plan)
+					continue
+				}
+			}
+		}
+
+	}
+	//TODO: need see if this go routine should be stopped and returned
```

Comment:
What work is left here?

Reply:
@djwhatle we probably need to find a way to pass the controller stop channel here so when all the controllers are stopped, we stop this watch source as well. Would like to take that as a follow on.

Reply:
I don't think we ever stop our controllers except for on termination of the process, but maybe you're thinking of some future use-case?